selfupdate.go

```diff
@@ -98,7 +105,13 @@ func selfupdate(ctx context.Context, gusCli *gusapi.APIClient, gusServer, machin
 		return fmt.Errorf("switching to non-active partition: %v", err)
 	}
 
-	log.Print("reboot")
+	log.Print("requesting reboot")
+	// TODO: change call from target.Reboot to something like target.ScheduleReboot
+	// which can asyncronouly perform the task instead of waiting
+	// for all the services to shutdown and then ack the reboot,
+	// othewise this causes a deadlock as the selfupdate service won't SIGTERM cleanly
+	// until the reboot ack is received, but won't receive the ack until all services shut down,
+	// causing a delayed reboot until SIGKILL kicks in.
 	if err := target.Reboot(); err != nil {
 		return fmt.Errorf("reboot: %v", err)
 	}
```

Comment:
This one will need some thought

Reply:
Hmm, why will selfupdate not SIGTERM cleanly in this state? I would have expected it to. SIGINT and SIGTERM are hooked up to context cancellation in selfupdate’s main function.